### PR TITLE
feat(document_symbol): added filtering

### PR DESCRIPTION
**Pull request by nhat-vo**
Implemented a generic filtering function to filter any tree, not limited to the filesystem tree as before. Currently only supports fzy. This enables filters in the document_symbol source. 

The behavior of this filtering is slightly different from the filtering of the filesystem source: pressing `<enter>` closes the input window but keeps the current filtered view, with the focus on the max-scored node. Calling `refresh` brings back the original view. I find this behavior more convenient as the user can now navigate in the filtered view, and exit it when they want.

Also added lua docstring to the document_symbol source.

### CODE REVIEW

**Review by cseickel — COMMENTED** [review]
It works well and I really appreciate you adding all the doc strings. One thing I would like to change is to make the filter consistent with the filesystem, which has both a `filter_on_submit` command (filter sticks after search closes) and `fuzzy_finder` command (you open selected item and clear search on enter). I appreciate that not everyone wants the same behavior which is why both options exist.

I personally only use the filter as a means of navigation, so to me having to hit enter once to close the filter, and then again to go to the symbol I searched for seems like a waste of effort. I also would not want to leave it filtered after finding the thing I wanted.

Can you add that form of searching and make it the binding for `/` so that the UX and command names are consistent between the sources? The current mode would be correct for the `f` binding.

File: lua/neo-tree/sources/document_symbols/lib/client_filters.lua [hunk 1]
```diff
@@ -28,18 +32,34 @@ local filter_clients = function(filter_type, filter_fn, resp)
   return result
 end
 
+---Filter only clients in the white list
```
[the remark this answers is not in the document]
Comment (nhat-vo) [reply lua/neo-tree/sources/document_symbols/lib/client_filters.lua]: Sure! I have renamed the options to `allow_only` and `ignore`.

**Comment by nhat-vo** [comment]
Sorry I forgot that `filter_on_submit` exists 😅. I have updated the behavior.

**Review by cseickel — APPROVED** [review]
Looks great!

**Pull request merged**